pkg/lib/infra/redis/hub.go

```diff
@@ -0,0 +1,244 @@
+)
+
+const (
+	RedisReadTimeout = 1 * time.Second
+)
+
+type SubscriberID struct {
+	Counter int64
+}
+
+func (i *SubscriberID) Get() int64 {
+	for {
+		curr := atomic.LoadInt64(&i.Counter)
+		if atomic.CompareAndSwapInt64(&i.Counter, curr, curr+1) {
+			return curr
+		}
+	}
+}
+
+type SubscriberKey struct {
+	ConnKey     string
+	ChannelName string
+}
+
+type Subscriber struct {
+	SubscriberID int64
+	Chan         chan *redis.Message
+}
+
+// Hub aims to multiplex multiple subscription to Redis PubSub channel over a single connection.
+type Hub struct {
+	Pool         *Pool
+	SubscriberID *SubscriberID
+	Logger       *log.Logger
+
+	mutex      sync.RWMutex
+	pubsub     map[string]*redis.PubSub
+	subscriber map[SubscriberKey][]Subscriber
+}
+
+func NewHub(pool *Pool, lf *log.Factory) *Hub {
+	h := &Hub{
+		Pool:         pool,
+		SubscriberID: &SubscriberID{},
+		Logger:       lf.New("redis-hub"),
+
+		subscriber: make(map[SubscriberKey][]Subscriber),
+		pubsub:     make(map[string]*redis.PubSub),
+	}
+	return h
+}
+
+func (h *Hub) Subscribe(
+	cfg *config.RedisConfig,
+	credentials *config.RedisCredentials,
+	channelName string,
+) (chan *redis.Message, func(), error) {
+	connKey := credentials.ConnKey()
+
+	h.mutex.Lock()
+	defer h.mutex.Unlock()
+
+	pubsub := h.getPubSub(cfg, credentials)
+	id := h.SubscriberID.Get()
+
+	key := SubscriberKey{
+		ConnKey:     connKey,
+		ChannelName: channelName,
+	}
+
+	c := make(chan *redis.Message)
+
+	h.subscriber[key] = append(h.subscriber[key], Subscriber{
+		SubscriberID: id,
+		Chan:         c,
+	})
+
+	// Subscribe if it is the first subscriber.
+	if len(h.subscriber[key]) == 1 {
+		h.Logger.Infof("subscribe because the first subscriber is joining")
+		ctx := context.Background()
+		err := pubsub.Subscribe(ctx, channelName)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	return c, func() {
+		h.cleanupOne(cfg, credentials, channelName, id)
+	}, nil
+}
+
+func (h *Hub) getPubSub(cfg *config.RedisConfig, credentials *config.RedisCredentials) *redis.PubSub {
+	connKey := credentials.ConnKey()
+
+	pubsub, ok := h.pubsub[connKey]
+	if ok {
+		return pubsub
+	}
+
+	ctx := context.Background()
+	client := h.Pool.Client(cfg, credentials)
+	pubsub = client.Subscribe(ctx)
+	h.pubsub[connKey] = pubsub
+
+	go func() {
+		for {
+			iface, err := pubsub.ReceiveTimeout(ctx, RedisReadTimeout)
+			if os.IsTimeout(err) {
+				continue
+			}
+			if err != nil {
+				h.cleanupAll(cfg, credentials)
+				return
+			}
+
+			switch n := iface.(type) {
+			case *redis.Subscription:
+				// Nothing special to do.
+				break
+			case *redis.Pong:
+				// Nothing special to do.
+				break
+			case *redis.Message:
+				key := SubscriberKey{
+					ConnKey:     connKey,
+					ChannelName: n.Channel,
+				}
+
+				h.mutex.RLock()
+				subs := make([]Subscriber, len(h.subscriber[key]))
+				copy(subs, h.subscriber[key])
+				h.mutex.RUnlock()
+
+				for _, s := range subs {
+					s.Chan <- n
```

Comment:
I think this may panic (i.e. sending to closed channel), since `cleanupOne` may close a channel in `subs` concurrently?

Reply:
Any idea on how to avoid that? Wrap the statement `s.Chan <- n` in another goroutine with recover() igoring the panic?

Reply:
We should close the channel when there is no more message to send to the channel (i.e. should close by sender). Therefore, I think we should close the channel in this goroutine instead.

Another thing I just noticed (by reading the doc of goredis), is that `pubsub.Channel()` is the preferred API. I think we can have a control channel (send from cleanup), and select on the control channel and message channel in this goroutine?